group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/ServerSideStickyRangeAssignor.java

```diff
@@ -0,0 +1,268 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * <p>The Server Side Sticky Range Assignor inherits properties of both the range assignor and the sticky assignor.
+ * Properties are :-
+ * <ul>
+ * <li> 1) Each consumer must get at least one partition per topic that it is subscribed to whenever the number of consumers is
+ *    less than or equal to the number of partitions for that topic. (Range) </li>
+ * <li> 2) Partitions should be assigned to consumers in a way that facilitates join operations where required. (Range) </li>
+ *    This can only be done if the topics are co-partitioned in the first place
+ *    Co-partitioned:-
+ *    Two streams are co-partitioned if the following conditions are met:-
+ * ->The keys must have the same schemas
+ * ->The topics involved must have the same number of partitions
+ * <li> 3) Consumers should retain as much as their previous assignment as possible. (Sticky) </li>
+ * </ul>
+ * </p>
+ *
+ * <p>The algorithm works mainly in 5 steps described below
+ * <ul>
+ * <li> 1) Get a map of the consumersPerTopic created using the member subscriptions.</li>
+ * <li> 2) Get a list of consumers (potentiallyUnfilled) that have not met the minimum required quota for assignment AND
+ * get a list of sticky partitions that we want to retain in the new assignment.</li>
+ * <li> 3) Add consumers from potentiallyUnfilled to Unfilled if they haven't met the total required quota = minQuota + (if necessary) extraPartition </li>
+ * <li> 4) Get a list of available partitions by calculating the difference between total partitions and assigned sticky partitions </li>
+ * <li> 5) Iterate through unfilled consumers and assign partitions from available partitions </li>
+ * </ul>
+ * </p>
+ *
+ */
+package org.apache.kafka.coordinator.group.assignor;
+
+
+import org.apache.kafka.common.Uuid;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.HashMap;
+import java.util.Collections;
+
+import static java.lang.Math.min;
+
+public class ServerSideStickyRangeAssignor implements PartitionAssignor {
+
+    public static final String RANGE_ASSIGNOR_NAME = "range-sticky";
+
+    @Override
+    public String name() {
+        return RANGE_ASSIGNOR_NAME;
+    }
+
+    protected static <K, V> void putList(Map<K, List<V>> map, K key, V value) {
```

Comment:
It's specific to the data structures we're using for the assignor so I made it protected so only classes that use the same Maps can use it, to avoid confusion.

Reply:
can we make these private?